tsfc/coffee_mode.py

```diff
@@ -221,28 +220,28 @@ def factorise_atomics(monomial_sum, optimal_atomics, argument_indices):
     """
     if not optimal_atomics:
         return monomial_sum
-    if len(monomial_sum.ordering) < 2:
+    if len(monomial_sum) < 2:
         return monomial_sum
     new_monomial_sum = MonomialSum()
     # Group monomials with respect to each optimal atomic
     factor_group = OrderedDict()
     for monomial in monomial_sum:
-        for sum_indices, oa in optimal_atomics:
-            if frozenset(monomial.sum_indices) == frozenset(sum_indices) and oa in monomial.atomics:
+        for oa in optimal_atomics:
+            if oa in monomial.atomics:
                 # Add monomial to the list of corresponding optimal atomic
-                factor_group.setdefault((sum_indices, oa), []).append(monomial)
+                factor_group.setdefault(oa, []).append(monomial)
                 break
         else:
             # Add monomials that do no have argument factors to new MonomialSum
-            new_monomial_sum.add(monomial.sum_indices, monomial.atomics, monomial.rest)
+            new_monomial_sum.add(*monomial)
```

Comment:
Ah, I think I start to understand what is going on here! It is the grouping pattern again! For the third time in this single pull request! Which means we must write a generic higher-order function to capture that pattern, just like a real Haskell programmer would do!

So again, you are grouping monomials, but this time the grouping key is the `set.intersection` with `optimal_atomics`. (Which you seem to assume to contain one or zero entries.)